### PR TITLE
Vehicles - Add Auto-Throttle to Planes

### PR DESCRIPTION
**Pull request by TheCandianVendingMachine**
**When merged this pull request will:**
- Add a speed-limiter type control to planes

Uses a basic PID Controller to control the throttle of a plane to keep it at some speed. PID was tuned in Simulink, so values work well. Once you have speed limiter on, you don't have to touch the throttle again.

[Video of it in action](https://youtu.be/cWYr35_laGw)

### CODE REVIEW

**Comment by PabstMirror** [comment]
Close https://github.com/acemod/ACE3/issues/2818 (I think)

**Comment by jonpas** [comment]
Why only planes? Most cars today have cruise control / tempomat. I assume the PID tune would be different though, so if you don't want to do that, that's fine. :D

**Review by jonpas — APPROVED** [review]

**Comment by PabstMirror** [comment]
ground vics are getting in-engine cruise in 2.06
https://community.bistudio.com/wiki/setCruiseControl

**Comment by jonpas** [comment]
Ah right! We have another PR somewhere for that already as well.

**Comment by TheCandianVendingMachine** [comment]
I don't think there is a way to control the throttle in non-plane vehicles, so I'd have to go into the config and `setVelocity` appropriately and it just turns into a mess

**Comment by jonpas** [comment]
[`setCruiseControl`](https://community.bistudio.com/wiki/setCruiseControl) has `autoThrust`.

**Review by PabstMirror — APPROVED** [review]

File: addons/vehicles/functions/fnc_autoThrottle.sqf [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        _throttle = 0 max ((_p + _d) min 1);
+    };
+
+    _vehicle setAirplaneThrottle _throttle;
+
+    _autothrottleParameters set [0, _d];
+    _autothrottleParameters set [1, _i];
+    _autothrottleParameters set [2, CBA_missionTime];
+    _autothrottleParameters set [3, _throttle];
+    _args set [2, _autothrottleParameters];
```
Comment (PabstMirror) [inline addons/vehicles/functions/fnc_autoThrottle.sqf:96]:
```suggestion
```

File: addons/vehicles/functions/fnc_autoThrottle.sqf [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    if (_driver != driver _vehicle) then {
+        GVAR(isSpeedLimiter) = false;
+    };
+
+    if (_throttleLogValue == 0) then {
+        _throttleLogValue = 1;
+    };
+
+    private _currentThrottle = (airplaneThrottle _vehicle) ^ _throttleLogValue;
+    if (_lastThrottleValue != -1 && EPSILON < abs (_currentThrottle - _lastThrottleValue)) then {
```
Comment (PabstMirror) [inline addons/vehicles/functions/fnc_autoThrottle.sqf:55]:
```suggestion
    if (_lastThrottleValue != -1 && {EPSILON < abs (_currentThrottle - _lastThrottleValue)}) then {
```

**Comment by dedmen** [comment]
Somewhat unrelated to this PR
> PID was tuned in Simulink

Engine setCruiseControl has hardcoded PID params that might not be optimal
kp = 0.5
ki = 0.3
kd = 0.2
maybe that can be tuned better.
If you can optimize that and find better values, feel free to contact me and i can transfer them into the engine vehicle PID, but it needs to work with all vehicles from quads to tanks

**Pull request merged**